022-custom-smtp-and-email-templates.md

```diff
@@ -0,0 +1,147 @@
+
+## Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+**What problem are you trying to solve?**
+
+Right now the whole Appwrite system uses a single global SMTP configuration set via environment variables, this means all the projects will use the same email configuration. As well there is no way for user to customize the default email templates. As Appwrite supports multiple projects, which may have different needs for emailing. It is even more crucial in the cloud to support custom email templating as each individual users will have their own need for the project and may want to use their own emailing system to send emails.
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+### API Endpoints
+
+For this we will introduce various endpoints
+
+**PATCH /v1/project/:projectId/smtp**       - an endpoint for updating smtp configuration for the project
+
+It will accept following parameters
+
+- **enabled** - boolean - **required** - whether or not the SMTP configuration is enabled
+- **sender** - string - SMTP sender email
+- **host** - string - SMTP server host name address
+- **port** - string - SMTP server tcp port
+- **secure** - string - SMTP secure protocol (empty or `tls`)
+- **username** - string - SMTP server user name
+- **password** - string - SMTP server password
+
+**PATCH /v1/project/:projectId/template** - an endpoint to update custom templates (email/sms)
+
+- **locale** - string - required (eg: en_us) - locale of the template
+- **key** - string - template key (email.emailVerification, sms.pincode) - the key that defines what the template is for
+- **template** - json encoded string - template content, that may have `subject`, `body`, `title`, `senderName`, `senderEmail`.
+
+**GET /v1/project/:projectId/template** - endpoint to get existing templates. accepts following parameters
+- **key** - string - key to identify the template
+- **locale** - string - locale of the template
+
+If custom template doesn't exist for the given details, the default server template will be returned.
+
+### Data Structure
+
+Project document will be updated with new attributes to support SMTP and templates configuration
+
+- **smtp**: json - save SMTP configuration for the project will have following fields
+    - **enabled** - boolean - whether or not the SMTP configuration is enabled
+    - **sender** - email of the sender
+    - **host** - SMTP server host name address
+    - **port** - SMTP server tcp port
+    - **secure** - SMTP secure protocol (empty or `tls`)
+    - **username** - SMTP server user name
+    - **password** - SMTP server password
+
+- **templates**: json - save custom templates
+    - **key** - string - key that defines what the template is for
```

Comment:
Should they key be a combination of the template type and locale? For example `verification-en_us`?

Reply:
may be `sms-verification-en_us` or `email-verification-en_us` ? including the channel?

Reply:
I think it can work